sql: changed LeaseManager API to return/use table descriptors

vivekmenezes opened the pull request:
The Acquire/AcquireByName and Release() methods no longer
return/use a LeaseState and instead return/use a
sqlbase.TableDescriptor with an expiration time.

the LeaseManager is moving towards a model where it returns
a table descriptor with an expiration time for the use of the
table descriptor. For now this table descriptor is associated
with a lease, but in the future it might be an older version of
the table descriptor, in which case it will not have a lease.

With this change, leases on the same table descriptor version
are renewed under the covers without needing the API to issue
an explicit release on a table descriptor version. The refcount
on the old lease is simply transferred to the new one. The new
lease is guaranteed to have a higher expiration time and thus a
prior returned expiration time returned by the API is guaranteed
to be contained within the new lease expiration time.

A Release() on the table descriptor will reduce the refcount on the
current lease. In the future, this will reduce the refcount on a table
version which might not be associated with a lease.

Removed the need to use removeLeaseIfExpiring since once a transaction
has picked a timestamp and a valid table descriptor for the timestamp
the table descriptor need not be reacquired through the API.

related to #2948

vivekmenezes committed:
sql: changed LeaseManager API to return/use table descriptors

The Acquire/AcquireByName and Release() methods no longer
return/use a LeaseState and instead return/use a
sqlbase.TableDescriptor with an expiration time.

the LeaseManager is moving towards a model where it returns
a table descriptor with an expiration time for the use of the
table descriptor. For now this table descriptor is associated
with a lease, but in the future it might be an older version of
the table descriptor, in which case it will not have a lease.

With this change, leases on the same table descriptor version
are renewed under the covers without needing the API to issue
an explicit release on a table descriptor version. The refcount
on the old lease is simply transferred to the new one. The new
lease is guaranteed to have a higher expiration time and thus a
prior returned expiration time returned by the API is guaranteed
to be contained within the new lease expiration time.

A Release() on the table descriptor will reduce the refcount on the
current lease. In the future, this will reduce the refcount on a table
version which might not be associated with a lease.

Removed the need to use removeLeaseIfExpiring since once a transaction
has picked a timestamp and a valid table descriptor for the timestamp
the table descriptor need not be reacquired through the API.

related to #2948

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16453)
<!-- Reviewable:end -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but please see the suggestion about thinning the responsibilities of `TableCollection`

I really like the direction here :D

---

Review status: 0 of 22 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/sql/lease.go, line 650 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgJYOjM-qGMhSWYHaw:-KmgJYOjM-qGMhSWYHax:b2fe621) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/lease.go#L650)):*
> ```Go
> 	s.mu.Lock()
> 	lease.mu.Lock()
> 	// subsume the refcount of the older leasse.
> ```

s/leasse/lease

---

*[pkg/sql/lease.go, line 981 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgMlrKUilWpNp-H-qa:-KmgMlrKUilWpNp-H-qb:b-mu8cqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/lease.go#L981)):*
> ```Go
> }
> 
> // LeaseManager manages acquiring and releasing per-table leases. It also
> ```

consider hanging a comment here about how clients get descriptors and expiration times, and how those expiration times should be compared with transaction timestamps

---

*[pkg/sql/lease.go, line 1102 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgLsLhJIBEEZfVgjc0:-KmgLsLhJIBEEZfVgjc1:b-qcaim9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/lease.go#L1102)):*
> ```Go
> 		// resolve the current or the old name.
> 		//
> 		// TODO(vivek): check if the entire above comment is indeed true.
> ```

I agree that revisiting the thinking around table renames may be a very good thing :)

---

*[pkg/sql/lease.go, line 1141 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgMVLCqV8SbibwuXFf:-KmgMVLCqV8SbibwuXFg:b-3arm8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/lease.go#L1141)):*
> ```Go
> // Acquire acquires a read lease for the specified table ID. If version is
> // non-zero the lease is grabbed for the specified version. Otherwise it is
> // grabbed for the most recent version of the descriptor that the lease manager
> ```

the comment should say something about the returned timestamp now

---

*[pkg/sql/lease.go, line 1163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgM8f31Gnc-blb5Q68:-KmgM8f31Gnc-blb5Q69:bkgb1p3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/lease.go#L1163)):*
> ```Go
> // the time of the call (i.e. if we were in the process of acquiring a lease
> // already, that lease is not good enough).
> // The returned lease had its refcount incremented, so the caller is responsible
> ```

this comment is now stale

---

*[pkg/sql/lease.go, line 1298 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgULaUB8-DZKYjRvNT:-KmgULaUB8-DZKYjRvNU:b-6cv8vm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/lease.go#L1298)):*
> ```Go
> // TableCollection is a collection of tables held by a single session that
> // serves SQL requests, or a background job using a table descriptor.
> type TableCollection struct {
> ```

I think this should move to `table.go`, where its methods are. No?

---

*[pkg/sql/lease.go, line 1302 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgOGMLDOoFRc_bVk16:-KmgOGMLDOoFRc_bVk17:b2tz183) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/lease.go#L1302)):*
> ```Go
> 	// A collection of table descriptor valid for the timestamp.
> 	// They are released once the transaction using them is complete.
> 	// If the transaction gets pushed and the timestamp changes,
> ```

I mostly like this comment (and the new structure of this struct), but here you talk about "the transaction", although no transaction has been introduced before. The struct comment says something about a "sql session". Which one is it? Is an instance of `TableCollection` used per session or per transaction?
I believe it may be a mixture of the two... The instance is per session, but some operations pass in a transaction? I think it'd help if the struct comment gave some guidance about how this is to be used (or if it described how it is used). 

And also consider transforming it into an instance per transaction (i.e. hold it in `TxnState`, not in `Session`). I'm thinking it may simplify how the state of the `TableCollection` is reset - the reset could be done externally, for example in `executor.go:runTxnAttempt()`.
Generally, I suspect it might be cleaner to make this struct oblivious to transactions and let its client to more. So move the resetting externally, and also don't have the `TableCollection` call `txn.UpdateDeadlineMaybe`, but have the client do that based on expiration timestamp returned by the `TableCollection`. There's only a handful of calls to `tc.getTableVersion()`, so I think that'd be feasible.
And then `TableCollection.timestamp` can also go.

---

*[pkg/sql/session.go, line 497 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgPZc6___bePlAQLs5:-KmgPZc7lpDvhN6UM5xQ:b-5pykxp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/session.go#L497)):*
> ```Go
> 	}
> 
> 	// Cleanup leases. We might have unreleased leases if we're finishing the
> ```

Change this comment from "cleanup leases" to... "decrement the refcount on descriptors"?
And perhaps you can take a look at what the deal with #7648 is :). I've just assigned it to you; it may be the case that it's obsolete.

---

*[pkg/sql/table.go, line 258 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgNUnxMZMN0ZrnEOad:-KmgNUnxMZMN0ZrnEOae:b9tbyop) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/table.go#L258)):*
> ```Go
> // table.ModificationTime <= txn.Timestamp < expirationTime
> //
> // TODO(vivek): Rollback most of #6418.
> ```

perhaps explain a bit what the deal with this 6418 is, if you're going to leave this comment in the code

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-:-KmgVQxG0fCw6nCplb1D:b-8kqdt5)*
<!-- Sent from Reviewable.io -->

vivekmenezes committed:
sql: move all TableCollection code to table.go

vivekmenezes commented:
Review status: 0 of 22 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/lease.go, line 650 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgJYOjM-qGMhSWYHaw:-Kmlp4b2BQUUPIdB8yIU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/lease.go#L650)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/leasse/lease
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 981 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgMlrKUilWpNp-H-qa:-Kmlvzey2RvcQmdXDZgu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/lease.go#L981)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider hanging a comment here about how clients get descriptors and expiration times, and how those expiration times should be compared with transaction timestamps
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 1102 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgLsLhJIBEEZfVgjc0:-KmlpCWCKTAcYz4m6Bxq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/lease.go#L1102)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I agree that revisiting the thinking around table renames may be a very good thing :)
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 1141 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgMVLCqV8SbibwuXFf:-KmlqD3sv6dEGneeGO6g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/lease.go#L1141)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

the comment should say something about the returned timestamp now
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 1163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgM8f31Gnc-blb5Q68:-Kmlr6aSzUmydT6LPS_d:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/lease.go#L1163)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this comment is now stale
</blockquote></details>

removed

---

*[pkg/sql/lease.go, line 1298 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgULaUB8-DZKYjRvNT:-KmlwrafcHu8MRFAVijl:b-jobmov) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/lease.go#L1298)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think this should move to `table.go`, where its methods are. No?
</blockquote></details>

added a new commit that moves TableCollection related stuff into table.go

---

*[pkg/sql/lease.go, line 1302 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgOGMLDOoFRc_bVk16:-KmlrhHbDRye6bSxoiUs:bx5xemx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/lease.go#L1302)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I mostly like this comment (and the new structure of this struct), but here you talk about "the transaction", although no transaction has been introduced before. The struct comment says something about a "sql session". Which one is it? Is an instance of `TableCollection` used per session or per transaction?
I believe it may be a mixture of the two... The instance is per session, but some operations pass in a transaction? I think it'd help if the struct comment gave some guidance about how this is to be used (or if it described how it is used). 

And also consider transforming it into an instance per transaction (i.e. hold it in `TxnState`, not in `Session`). I'm thinking it may simplify how the state of the `TableCollection` is reset - the reset could be done externally, for example in `executor.go:runTxnAttempt()`.
Generally, I suspect it might be cleaner to make this struct oblivious to transactions and let its client to more. So move the resetting externally, and also don't have the `TableCollection` call `txn.UpdateDeadlineMaybe`, but have the client do that based on expiration timestamp returned by the `TableCollection`. There's only a handful of calls to `tc.getTableVersion()`, so I think that'd be feasible.
And then `TableCollection.timestamp` can also go.
</blockquote></details>

I agree there is more that we need to do here. In general once I've gotten this and few more PRs in, including the one that moves the lease manager to use its own transactions (the original PR that I was working on), we'll have a more flexible environment to move things to the right place and not pass transactions into this code.

I'd also like for prepared statements to use the API and keep reevaluating themselves once they hit the expiration time of their tables

---

*[pkg/sql/session.go, line 497 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgPZc6___bePlAQLs5:-KmltRXzj88jD8GNwG_n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/session.go#L497)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Change this comment from "cleanup leases" to... "decrement the refcount on descriptors"?
And perhaps you can take a look at what the deal with #7648 is :). I've just assigned it to you; it may be the case that it's obsolete.
</blockquote></details>

Done.

---

*[pkg/sql/table.go, line 258 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453#-KmgNUnxMZMN0ZrnEOad:-KmluVr9c_K4qw84wqol:byn1gr4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95331aa8b35a371fad40c16ae03db71ef22fdef/pkg/sql/table.go#L258)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

perhaps explain a bit what the deal with this 6418 is, if you're going to leave this comment in the code
</blockquote></details>

Added a comment explaining

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16453)*
<!-- Sent from Reviewable.io -->
